### PR TITLE
Do not add into vectors if they do not exist

### LINKED ISSUE #11705: Having a nodal constraint on a displaced mesh results in a segfault
## Rationale

Trying to solve a problem with a nodal constraint on a displaced mesh results in a segfault.

## Description

This input file reproduces the problem:
```
[Mesh]
  file = 2-lines.e
  displacements = 'disp_x'
[]

[AuxVariables]
  [./disp_x]
  [../]
[]

[AuxKernels]
  [./disp_x_ak]
    type = ConstantAux
    variable = disp_x
    value = 1
  [../]
[]

[Variables]
  [./u]
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]
[]

[BCs]
  [./left]
    type = DirichletBC
    variable = u
    boundary = 1
    value = 1
  [../]

  [./right]
    type = DirichletBC
    variable = u
    boundary = 4
    value = 3
  [../]
[]

[Constraints]
  [./c1]
    type = EqualValueNodalConstraint
    variable = u
    master = 0
    slave = 4
    penalty = 100000
    use_displaced_mesh = true
  [../]
[]

[Executioner]
  type = Steady

  solve_type = 'PJFNK'
[]

[Outputs]
  exodus = true
[]
```

## Impact

Unable to solve problems with nodal constraints on displaced meshes.

### PR DESCRIPTION
**Pull request by andrsd**
Closes #11705

### CODE REVIEW

File: framework/src/problems/DisplacedProblem.C [hunk 1]
```diff
@@ -676,9 +676,10 @@ DisplacedProblem::addCachedResidual(THREAD_ID tid)
...
 DisplacedProblem::addCachedResidualDirectly(NumericVector<Number> & residual, THREAD_ID tid)
 {
-
-  _assembly[tid]->addCachedResidual(residual, _displaced_nl.timeVectorTag());
-  _assembly[tid]->addCachedResidual(residual, _displaced_nl.nonTimeVectorTag());
+  if (_displaced_nl.hasVector(_displaced_nl.timeVectorTag()))
+    _assembly[tid]->addCachedResidual(residual, _displaced_nl.timeVectorTag());
+  if (_displaced_nl.hasVector(_displaced_nl.nonTimeVectorTag()))
+    _assembly[tid]->addCachedResidual(residual, _displaced_nl.nonTimeVectorTag());
 }
```
Comment (fdkong) [inline framework/src/problems/DisplacedProblem.C]: I already check this in` Assembly.C`

```
void
Assembly::addCachedResidual(NumericVector<Number> & residual, TagID tag_id)
{
  if (!_sys.hasVector(tag_id))
  {
    _cached_residual_values[tag_id].clear();
    _cached_residual_rows[tag_id].clear();
    return;
  }
```

Does not this work for you? What is the issue here I am missing.

**Review by fdkong — DISMISSED** [review]

File: framework/src/problems/DisplacedProblem.C (hunk 1 above)
Comment (andrsd) [reply framework/src/problems/DisplacedProblem.C]: Also, check `void
FEProblemBase::addCachedResidualDirectly(NumericVector<Number> & residual, THREAD_ID tid)`. There we do:
```
  if (_nl->hasVector(_nl->timeVectorTag()))
    _assembly[tid]->addCachedResidual(residual, _nl->timeVectorTag());
  if (_nl->hasVector(_nl->nonTimeVectorTag()))
    _assembly[tid]->addCachedResidual(residual, _nl->nonTimeVectorTag());
```

So I mimicked that in the `DisplacedProblem`.

Comment (fdkong) [reply framework/src/problems/DisplacedProblem.C]: I see. I will check why Assembly does not take care of this.

**Comment by fdkong** [comment]
Are you fine with this fix? @andrsd

**Comment by andrsd** [comment]
> Are you fine with this fix?

Yes. Looks good.

**Review by permcody — APPROVED** [review]

**Pull request merged**